### ISSUE TITLE
FilePath.copyTo raise OSError if source doesn't exist

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5017]
|<img alt="techtonik's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| techtonik reported|
|-|-|
|Trac ID|trac#5017|
|Type|defect|
|Created|2011-04-03 17:28:49Z|
|Branch|https://github.com/twisted/twisted/tree/copyto-missing-5017|

Attempt to copy file/dir that doesn't exist, results in NotImplementedError exception:
```
...
  File "C:\Python27\lib\site-packages\twisted\python\filepath.py", line 900, in copyTo
    "Only copying of files and directories supported")
NotImplementedError: Only copying of files and directories supported
```
How about raising OSError with ENOENT code instead?

Attachments:

* [issue5017.filepath.copyTo.nonexistent.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d3c/d3ca45c909afe56d2c4ecb58e7b4541381084dbd/3ee9eb0e0e71957ca63334acbbcb5b8a52a5962f.patch) (951 bytes) - added by techtonik on 2011-04-03 17:29:48Z - 
* [5017.filepath.copyTo.nonexistent.with.test.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d3c/d3ca45c909afe56d2c4ecb58e7b4541381084dbd/f1e2eeaeb295fd22875c0de6ba61c2a4a62fa459.patch) (1839 bytes) - added by techtonik on 2011-04-04 07:07:11Z - patch with test

<details><summary>Searchable metadata</summary>

```
trac-id__5017 5017
type__defect defect
reporter__techtonik techtonik
priority__normal normal
milestone__ 
branch__branches_copyto_missing_5017 branches/copyto-missing-5017
branch_author__techtonik techtonik
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1301851729000000 1301851729000000
changetime__1302551755000000 1302551755000000
version__None None
owner__ 
cc__techtonik
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

Woops.  Yes, that seems like a good idea.

**Comment by twisted-trac** [comment]
|<img alt="techtonik's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|techtonik commented|
|-|-|

Wow, that was fast! Did you catch the patch?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to techtonik|
|-|-|

> Wow, that was fast! Did you catch the patch?

Nope, just trying to give feedback about the ticket itself. :)

The patch looks like a good start.  Can you also add a unit test which exercises this case?

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="techtonik's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|techtonik commented|
|-|-|

The trick is to learn how to deal with trial and ensure it is running tests using code base from working copy.

**Comment by twisted-trac** [comment]
|<img alt="techtonik's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|techtonik removed owner|
|-|-|

Updated patch includes exercising testcase.

**Comment by twisted-trac** [comment]
|<img alt="techtonik's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|techtonik commented|
|-|-|

(as glyph pointed out in comment:2:ticket:5019 `easy` keyword is only for new contributors to find the stuff to work on - not for review requests to core developers)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [31639]) Branching to 'copyto-missing-5017'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [31640]) Apply patch from the ticket with minor tweaks

refs [#5017](https://github.com/twisted/twisted/issues/5017)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

Thanks.

  1. I moved the tests into `twisted/test/test_paths.py` (which is not the place I would expect you to look to find tests for this module, so I'm not surprised you didn't find it - however, in the future, there's another hint: the `test-case-name` value at the top of some files.  `filepath.py` has one that points at the right place).
  1. I also changed the test to use `self.mktemp())` as the source path, since that's more or less guaranteed not to exist.
  1. I also think the `OSError` string message should try to match what the platform would actually give you.  Doing this totally correctly probably involves calling `os.strerror`, but it's probably fine to just pass in "No such file or directory" for starters.

I checked the patch into a branch and made these changes.  If the [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/copyto-missing-5017) look good, I'll merge it.  Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set status to `closed`|
|-|-|

(In [31643]) Merge copyto-missing-5017

Author: techtonik
Reviewer: exarkun
Fixes: [#5017](https://github.com/twisted/twisted/issues/5017)

Raise `OSError(ENOENT)` if an attempt is made to copy a file which does
not exist using `FilePath.copyTo`.
